static/templates/recipient_row.hbs

```diff
@@ -27,7 +27,7 @@
             <span class="stream_topic">
                 {{! topic link }}
                 <a class="message_label_clickable narrows_by_topic"
-                  href="{{topic_url}}"
+                  href="{{#if collapse_messages}}{{topic_url}}{{else}}{{message_containers.0.msg.url}}{{/if}}"
```

Comment:
It's probably better to do the conditional/logic part in `message_list_view.js` somewhere rather than here; we try to avoid having conditionals in templates because it makes them hard to find.

Also, that'll make it easier to add the block comment we'll want explaining the intent of this feature, which will be useful reading.